webapp/lib/controller.dart

```diff
@@ -267,48 +277,86 @@ bool isProjectStale(model.NeedsReplyData projectData) {
   }
 }
 
-void setupProjectTimer(model.NeedsReplyData projectData, [bool stale = false]) {
-  projectTimers[selectedProject]?.cancel();
+String getWatchdogTimer(Object data) {
+  if (data is model.NeedsReplyData) {
+    return  selectedProject;
+  } else if (data is model.SystemMetricsData) {
+    return SYSTEM_METRICS_ROOT_COLLECTION_KEY;
+  } else {
+    throw new model.ModelDoesNotExistException('Model does not exist');
+  }
+}
+
+void setupWatchdogTimer<T>(T latestData, [bool stale = false]) {
+  var data;
+  if (latestData is model.NeedsReplyData) {
+    data = latestData as model.NeedsReplyData;
+  } else if (latestData is model.SystemMetricsData) {
+    data = latestData as model.SystemMetricsData;
+  }
+
+  watchdogTimers[getWatchdogTimer(data)]?.cancel();
 
   if (stale) {
-    projectTimers[selectedProject] = null;
+    watchdogTimers[getWatchdogTimer(data)] = null;
   } else {
-    var timeToExecute =  projectData.datetime.add(Duration(hours: 2));
+    var timeToExecute =  data.datetime.add(Duration(hours: 2));
     var now = new DateTime.now();
     var duration = timeToExecute.difference(now);
     var timer = new Timer(duration, () {
-      view.contentView.stale = true;
+      if (data?.project == selectedProject) {
+        view.contentView.setStale(NEEDS_REPLY_METRICS_COLLECTION_KEY, true);
+      }
+      view.contentView.setStale(SYSTEM_METRICS_ROOT_COLLECTION_KEY, true);
     });
-    projectTimers[selectedProject] = timer;
+    watchdogTimers[getWatchdogTimer(data)] = timer;
   }
 }
 
 void checkNeedsReplyMetricsStale(List<model.NeedsReplyData> updatedData) {
   if (updatedData.isEmpty) {
-    var selectedProjectTimer = projectTimers[selectedProject];
-    selectedProjectTimer?.cancel();
-    view.contentView.stale = false;
+    var selectedTimer = watchdogTimers[selectedProject];
+    selectedTimer?.cancel();
+    view.contentView.setStale(NEEDS_REPLY_METRICS_COLLECTION_KEY, false);
     return;
   }
 
   updatedData.sort((d1, d2) => d1.datetime.compareTo(d2.datetime));
 
-  var latestProjectData = updatedData.last ?? null;
+  var latestData = updatedData.last ?? null;
 
-  if (latestProjectData != null) {
+  if (latestData != null) {
```

Comment:
as per comment above, I think `latestData` really shouldn't be null. So what I suggest we do instead is: if it's null, then throw an error and show it via the snackbar. if it's not null, we call `setupWatchDogTimer`.

Unless I'm misunderstanding the cases where `latestData` can be correctly `null`?

Reply:
Now doing the data empty check before all this as we discussed.